services/venice-controller/src/main/java/com/linkedin/venice/controller/systemstore/SystemStoreRepairTask.java

```diff
@@ -0,0 +1,355 @@
+              retrievedHeartbeatTimestamp);
+        }
+      }
+    }
+  }
+
+  void updateBadSystemStoreCount(String clusterName, Set<String> newUnhealthySystemStoreSet) {
+    long newBadMetaSystemStoreCount = newUnhealthySystemStoreSet.stream()
+        .filter(x -> VeniceSystemStoreType.getSystemStoreType(x).equals(VeniceSystemStoreType.META_STORE))
+        .count();
+    getBadMetaStoreCount(clusterName).set(newBadMetaSystemStoreCount);
+    getBadPushStatusStoreCount(clusterName).set(newUnhealthySystemStoreSet.size() - newBadMetaSystemStoreCount);
+    LOGGER.info(
+        "Cluster: {} collected unhealthy system stores: {}. Meta system store count: {}, push status system store count: {}",
+        clusterName,
+        newUnhealthySystemStoreSet.toString(),
+        getBadMetaStoreCount(clusterName).get(),
+        getBadPushStatusStoreCount(clusterName).get());
+  }
+
+  void updateUnreachableSystemStoreCount(String clusterName, Set<String> unreachableSystemStoreSet) {
+    getUnreachableSystemStoreCounter(clusterName).set(unreachableSystemStoreSet.size());
+    LOGGER.info("Cluster: {} has unreachable system stores: {}", clusterName, unreachableSystemStoreSet);
+  }
+
+  void updateNotRepairableSystemStoreCount(String clusterName, Set<String> notRepairableSystemStoreSet) {
+    getNotRepairableSystemStoreCounter(clusterName).set(notRepairableSystemStoreSet.size());
+    LOGGER.info("Cluster: {} has not repairable system stores: {}", clusterName, notRepairableSystemStoreSet);
+  }
+
+  AtomicBoolean getIsRunning() {
+    return isRunning;
+  }
+
+  public Map<String, ControllerClient> getControllerClientMap(String clusterName) {
+    return getParentAdmin().getVeniceHelixAdmin().getControllerClientMap(clusterName);
+  }
+
+  void sendHeartbeatToSystemStore(String clusterName, String systemStoreName, long heartbeatTimestamp) {
+    for (Map.Entry<String, ControllerClient> entry: getControllerClientMap(clusterName).entrySet()) {
+      entry.getValue().sendHeartbeatToSystemStore(systemStoreName, heartbeatTimestamp);
+    }
+  }
+
+  long getHeartbeatFromSystemStore(String clusterName, String systemStoreName) {
+    long oldestHeartbeatTimestamp = Long.MAX_VALUE;
+    for (Map.Entry<String, ControllerClient> entry: getControllerClientMap(clusterName).entrySet()) {
+      oldestHeartbeatTimestamp = Math.min(
+          oldestHeartbeatTimestamp,
+          entry.getValue().getHeartbeatFromSystemStore(systemStoreName).getHeartbeatTimestamp());
+    }
+    return oldestHeartbeatTimestamp;
+  }
+
+  /**
+   * Poll the system store push status until it reaches terminal status.
+   * Based on system store setup, stuck job should time out in 1h.
+   * If push job completes, it will return true, otherwise return false.
+   */
+  boolean pollSystemStorePushStatusUntilCompleted(String clusterName, String systemStoreName, int version) {
+    String kafkaTopic = Version.composeKafkaTopic(systemStoreName, version);
+    while (true) {
```

Comment:
Shall we put a upper limit here otherwise this poll function can stuck forever in some scenario?

Reply:
My understanding is if it is stuck, after one hour it should timeout right? Or are you suggesting we add extra safeguard here?

Reply:
Let us keep this for now.